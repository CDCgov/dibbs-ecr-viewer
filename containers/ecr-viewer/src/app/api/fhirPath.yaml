patientNameList: "Bundle.entry.resource.where(resourceType = 'Patient').name"
patientAddressList: "Bundle.entry.resource.where(resourceType = 'Patient').address"
patientTelecom: "Bundle.entry.resource.where(resourceType = 'Patient').telecom"
patientCounty: "Bundle.entry.resource.where(resourceType = 'Patient').address.first().county"
patientCountry: "Bundle.entry.resource.where(resourceType = 'Patient').address.first().country"

patientIds: "Bundle.entry.resource.where(resourceType = 'Patient').identifier.where(system != 'urn:ietf:rfc:3986').value.join('\n')"
patientDOB: "Bundle.entry.resource.where(resourceType = 'Patient').birthDate"
patientVitalStatus: "Bundle.entry.resource.where(resourceType = 'Patient').deceasedBoolean"
patientDOD: "Bundle.entry.resource.where(resourceType = 'Patient').deceasedDate"
patientGender: "Bundle.entry.resource.where(resourceType = 'Patient').gender"
patientRace: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.where(url = 'ombCategory').value"
patientRaceDetailed: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension.where(url = 'detailed').value"
patientEthnicity: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.where(url = 'ombCategory').value"
patientEthnicityDetailed: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url = 'http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity').extension.where(url = 'detailed').value"
patientCommunication: "Bundle.entry.resource.where(resourceType = 'Patient').communication"
patientTribalAffiliation: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url='http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-tribal-affiliation-extension').extension.where(url='TribeName').value"
patientEmergencyContact: "Bundle.entry.resource.where(resourceType = 'Patient').contact"

# Social History
patientCurrentJobTitle: "Bundle.entry.resource.where(resourceType='Observation').where(meta.profile='http://hl7.org/fhir/us/odh/StructureDefinition/odh-PastOrPresentJob').where(effectivePeriod.end.exists().not()).value"
patientTobaccoUse: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.code='72166-2').where(category.coding.code='social-history').value"
patientHomelessStatus: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.code='75274-1').where(category.coding.code='social-history').value"
patientPregnancyStatus: "Bundle.entry.resource.where(resourceType='Observation').where(meta.profile='http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-pregnancy-status-observation').value"
patientAlcoholUse: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.where(code='11331-6' and system = 'http://loinc.org')).value"
patientAlcoholIntake: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.where(code='74013-4' and system = 'http://loinc.org')).value"
patientAlcoholComment: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.display='Alcohol Comment').value"
patientSexualOrientation: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.code='76690-7').value"
patientGenderIdentity: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url='http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-genderidentity-extension').value"
patientReligion: "Bundle.entry.resource.where(resourceType = 'Patient').extension.where(url='http://hl7.org/fhir/StructureDefinition/patient-religion').value"
patientMaritalStatus: "Bundle.entry.resource.where(resourceType = 'Patient').maritalStatus"

#eCR Metadata
eicrIdentifier: "Bundle.entry.resource.where(resourceType= 'Composition').id"
eicrReleaseVersion: "Bundle.entry.resource.where(resourceType='Composition').extension.where(url='https://www.hl7.org/implement/standards/product_brief.cfm?product_id=436').valueString"
eicrCustodianRef: "Bundle.entry.resource.where(resourceType= 'Composition').custodian.reference"
dateTimeEcrCreated: "Bundle.entry.resource.where(resourceType = 'Composition').date"
ehrSoftware: "Bundle.entry.resource.where(resourceType = 'Device').where(property[0].type.coding.code='software').version.value"
ehrManufacturerModel: "Bundle.entry.resource.where(resourceType = 'Device').where(property[0].type.coding.code='software').manufacturer"
eRSDwarnings: "Bundle.entry.resource.where(resourceType= 'Composition').section.where(title = 'Reportability Response Information Section').extension.where(url = 'http://hl7.org/fhir/us/ecr/StructureDefinition/rr-eicr-processing-status-extension').valueCodeableConcept.coding.entries.eRSDwarnings"
compositionAuthorRefs: "Bundle.entry.resource.where(resourceType = 'Composition').author"

# Encounter Info
encounterEndDate: "Bundle.entry.resource.where(resourceType = 'Encounter').period.end"
encounterStartDate: "Bundle.entry.resource.where(resourceType = 'Encounter').period.start"
encounterDiagnosis: "Bundle.entry.resource.where(resourceType = 'Encounter').diagnosis"
encounterType: "Bundle.entry.resource.where(resourceType='Encounter')[0].class.display"
encounterID: "Bundle.entry.resource.where(resourceType='Encounter')[0].identifier"
facilityContact: "Bundle.entry.resource.where(resourceType = 'Location')[0].telecom.where(system = 'phone')[0].value"
facilityContactAddress: "Bundle.entry.resource.where(resourceType = 'Encounter')[0].serviceProvider"
facilityLocation: "Bundle.entry.resource.where(resourceType = 'Encounter')[0].location[0].location.reference"
facilityName: "Bundle.entry.resource.where(resourceType = 'Encounter')[0].location[0].location.display"
facilityAddress: "Bundle.entry.resource.where(resourceType = 'Location')[0].address"
facilityType: "Bundle.entry.resource.where(resourceType = 'Encounter')[0].location[0].extension.where(url = 'http://build.fhir.org/ig/HL7/case-reporting/StructureDefinition-us-ph-location-definitions.html#Location.type').value"

compositionEncounterRef: "Bundle.entry.resource.where(resourceType = 'Composition').encounter.reference"
encounterIndividualRef: "Encounter.participant.where(type.coding.code = 'ATND').individual.reference"
encounterParticipants: "Encounter.participant"

rrDetails: "Bundle.entry.resource.where(meta.profile = 'http://hl7.org/fhir/us/ecr/StructureDefinition/rr-reportability-information-observation')"

# Clinical Data
clinicalReasonForVisit: "Bundle.entry.resource.section.where(title.lower() = 'reason for visit')[0].extension[0].value"

# Vitals
patientHeight: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.code='8302-2').first().valueQuantity.value"
patientHeightMeasurement: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='8302-2').first().valueQuantity.unit"
patientHeightDate: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='8302-2').first().effectiveDateTime"
patientWeight: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='29463-7').first().valueQuantity.value"
patientWeightMeasurement: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='29463-7').first().valueQuantity.unit"
patientWeightDate: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='29463-7').first().effectiveDateTime"
patientBmi: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='39156-5').valueQuantity.value"
patientBmiMeasurement: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='39156-5').valueQuantity.unit"
patientBmiDate: "Bundle.entry.resource.where(resourceType='Observation').where(code.coding.first().code='39156-5').effectiveDateTime"

resolve: "Bundle.entry.resource.where(resourceType = %resourceType).where(id = %id)"

# Clinical Info
activeProblems: "Bundle.entry.resource.where(resourceType='Condition').where(category.coding.code='problem-item-list')"
activeProblemsDisplay: "Condition.code.coding.display.first()"
activeProblemsOnsetDate: "Condition.onsetDateTime"
activeProblemsOnsetAge: "Condition.onsetAge.value"
activeProblemsComments: "Condition.note[0].text"
historyOfPresentIllness: "Bundle.entry.resource.where(resourceType = 'Composition').section.where(code.coding.code = '10164-2').text.`div`.first()"

# Treatment Details
planOfTreatment: "Bundle.entry.resource.section.where(title = 'Plan of Treatment').text.first().`div`"
plannedProcedures: "Bundle.entry.resource.where(resourceType = 'CarePlan').activity"
plannedProcedureName: "detail.code.coding[0].display"
plannedProcedureOrderedDate: "extension.where(url = 'dibbs.orderedDate').valueString"
plannedProcedureScheduledDate: "detail.scheduledString"

# Administered Medications
adminMedicationsRefs: "Bundle.entry.resource.section.where(code.coding[0].code = '29549-3').entry.reference"

## Care Team
careTeamParticipants: "Bundle.entry.resource.where(resourceType='CareTeam').participant"
careTeamParticipantMemberName: "member.name"
careTeamParticipantRole: "role.text"
careTeamParticipantStatus: "modifierExtension.where(url='participant.status').valueString"
careTeamParticipantPeriod: "period.text"

# Immunization Info
immunizations: "Bundle.entry.resource.where(resourceType = 'Immunization')"
immunizationsName: "Immunization.vaccineCode.coding.display.first()"
immunizationsAdminDate: "Immunization.occurrenceDateTime"
immunizationsDoseNumber: "Immunization.protocolApplied.where(doseNumberPositiveInt.exists()).doseNumberPositiveInt"
immunizationsManufacturerName: "Immunization.manufacturer.name"
immunizationsLotNumber: "Immunization.lotNumber"

# Procedure
procedures: "Bundle.entry.resource.where(resourceType='Procedure')"
procedureName: "Procedure.code.coding.iif(where(system='http://loinc.org').display.exists(), where(system='http://loinc.org').display, display.first())"
procedureDate: "Procedure.performedDateTime"
procedureReason: "Procedure.reason.display"

# Lab Info
diagnosticReports: "Bundle.entry.resource.where(resourceType = 'DiagnosticReport')"
observations: "Bundle.entry.resource.where(resourceType = 'Observation')"
labResultDiv: "Bundle.entry.resource.section.where(code.coding[0].code = '30954-2').text.`div`.first()"
specimenCollectionTime: "Observation.extension[0].extension.where(url = 'specimen collection time').valueDateTime"
specimenReceivedTime: "Observation.extension[0].extension.where(url = 'specimen receive time').valueDateTime"
specimenSource: "Observation.extension[0].extension.where(url = 'specimen source').valueString"
observationReferenceValue: "Observation.extension[0].extension.where(url = 'observation entry reference value').valueString"
observationComponent: "code.coding.display.first()"
# observationValue logic: <value> <units> (<interpretation>)
observationValue: (valueQuantity.value.toString() | valueString | valueCodeableConcept.coding.display | iif(valueQuantity.unit.exists(), iif(valueQuantity.unit = '%', valueQuantity.unit, ' ' + valueQuantity.unit), '') | iif(interpretation.coding.display.exists(), ' (' + interpretation.coding.display + ')', '')).join('')
observationReferenceRange: "referenceRange.text"
observationDeviceReference: "device.reference"
observationNote: "note.text"
observationOrganism: "code.coding.display.first()"
observationAntibiotic: "code.coding.display.first()"
observationOrganismMethod: "extension.where(url = 'methodCode originalText').valueString"
observationSusceptibility: "valueString"

# Organization
organizations: "Bundle.entry.resource.where(resourceType = 'Organization')"

# Travel History
patientTravelHistory: "Bundle.entry.resource.where(resourceType='Observation').where(meta.profile='http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-travel-history')"
travelHistoryStartDate: "effectivePeriod.start"
travelHistoryEndDate: "effectivePeriod.end"
travelHistoryLocation: "component.where(code.coding.code='LOC').valueCodeableConcept.text"
travelHistoryPurpose: "component.where(code.coding.code='280147009').valueCodeableConcept.coding.display"

# Stamped
stampedImmunizations: "entry.resource.where(extension('https://reportstream.cdc.gov/fhir/StructureDefinition/condition-code').valueCoding.code = %snomedCode and resourceType = 'Immunization')"
