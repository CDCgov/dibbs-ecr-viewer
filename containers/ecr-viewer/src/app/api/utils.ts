import * as fs from "fs";
import * as path from "path";
import yaml from "js-yaml";
import { PathMappings } from "@/app/utils/data-utils";

export const S3_SOURCE = "s3";
export const AZURE_SOURCE = "azure";

/**
 * Loads the YAML configuration for path mappings from a predefined file location.
 * @returns An object representing the path mappings defined in the YAML configuration file.
 */
export function loadYamlConfig(): PathMappings {
  const filePath = path.join(process.cwd(), "src/app/api/fhirPath.yaml");
  const fileContents = fs.readFileSync(filePath, "utf8");
  return <PathMappings>yaml.load(fileContents);
}

/**
 * Converts stream data to json data
 * @param stream - The input stream that provides JSON data in chunks. The stream
 *   should implement the async iterable protocol to allow for-await-of
 *   iteration over its data chunks.
 * @returns A promise that resolves to the JSON-parsed object from the accumulated
 *  stream data. The specific structure of this object depends on the JSON
 *  content of the stream.
 */
export async function streamToJson(stream: NodeJS.ReadableStream | undefined) {
  if (!stream) throw new Error("Stream is undefined");

  let rawData = "";
  for await (const chunk of stream) {
    rawData += chunk;
  }
  return JSON.parse(rawData);
}
