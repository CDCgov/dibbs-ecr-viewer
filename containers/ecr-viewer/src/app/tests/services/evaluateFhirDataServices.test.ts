import { loadYamlConfig } from "@/app/api/utils";
import {
  evaluateEncounterId,
  evaluateFacilityId,
  evaluatePatientRace,
  evaluatePatientEthnicity,
  evaluatePractitionerRoleReference,
  evaluateReference,
  evaluateValue,
  evaluateEmergencyContact,
  evaluatePatientAddress,
  evaluatePatientName,
  evaluateDemographicsData,
  evaluateEncounterCareTeamTable,
  evaluateAlcoholUse,
  evaluatePatientLanguage,
  evaluatePatientVitalStatus,
  getHumanReadableCodeableConcept,
} from "@/app/services/evaluateFhirDataService";
import { Bundle, CodeableConcept, Patient } from "fhir/r4";
import BundleMiscNotes from "@/app/tests/assets/BundleMiscNotes.json";
import BundlePatient from "@/app/tests/assets/BundlePatient.json";
import BundleEcrMetadata from "@/app/tests/assets/BundleEcrMetadata.json";
import BundlePractitionerRole from "@/app/tests/assets/BundlePractitionerRole.json";
import BundlePatientMultiple from "@/app/tests/assets/BundlePatientMultiple.json";

const mappings = loadYamlConfig();

describe("evaluateFhirDataServices tests", () => {
  describe("Evaluate Reference", () => {
    it("should return undefined if resource not found", () => {
      const actual = evaluateReference(
        BundleMiscNotes as unknown as Bundle,
        mappings,
        "Observation/1234",
      );

      expect(actual).toBeUndefined();
    });
    it("should return the resource if the resource is available", () => {
      const actual = evaluateReference(
        BundlePatient as unknown as Bundle,
        mappings,
        "Patient/99999999-4p89-4b96-b6ab-c46406839cea",
      );

      expect(actual.id).toEqual("99999999-4p89-4b96-b6ab-c46406839cea");
      expect(actual.resourceType).toEqual("Patient");
    });
  });

  describe("evaluate value", () => {
    it("should provide the string in the case of valueString", () => {
      const actual = evaluateValue(
        { resourceType: "Observation", valueString: "abc" } as any,
        "value",
      );

      expect(actual).toEqual("abc");
    });
    it("should provide the string in the case of valueCodeableConcept", () => {
      const actual = evaluateValue(
        {
          resourceType: "Observation",
          valueCodeableConcept: {
            coding: [
              {
                display: "Negative",
                code: "N",
              },
            ],
          },
        } as any,
        "value",
      );

      expect(actual).toEqual("Negative");
    });
    it("should provide the string in the case of valueCoding", () => {
      const actual = evaluateValue(
        {
          resourceType: "Extension",
          valueCoding: {
            display: "Negative",
            code: "N",
          },
        } as any,
        "value",
      );

      expect(actual).toEqual("Negative");
    });
    it("should provide the string in the case of valueBoolean", () => {
      const actual = evaluateValue(
        {
          resourceType: "Extension",
          valueBoolean: true,
        } as any,
        "value",
      );

      expect(actual).toEqual("true");
    });
    it("should provide the code as a fallback in the case of valueCodeableConcept", () => {
      const actual = evaluateValue(
        {
          resourceType: "Observation",
          valueCodeableConcept: {
            coding: [
              {
                code: "N",
              },
            ],
          },
        } as any,
        "value",
      );

      expect(actual).toEqual("N");
    });
    it("should provide the code as a fallback in the case of valueCoding", () => {
      const actual = evaluateValue(
        {
          resourceType: "Extension",
          valueCoding: {
            code: "N",
          },
        } as any,
        "value",
      );

      expect(actual).toEqual("N");
    });
    describe("Quantity", () => {
      it("should provide the value and string unit with a space inbetween", () => {
        const actual = evaluateValue(
          {
            resourceType: "Observation",
            valueQuantity: { value: 1, unit: "ft" },
          } as any,
          "value",
        );

        expect(actual).toEqual("1 ft");
      });
      it("should provide the value and symbol unit", () => {
        const actual = evaluateValue(
          {
            resourceType: "Observation",
            valueQuantity: { value: 1, unit: "%" },
          } as any,
          "value",
        );

        expect(actual).toEqual("1%");
      });
    });
  });

  describe("Evaluate Identifier", () => {
    it("should return the Identifier value", () => {
      const actual = evaluateValue(
        BundlePatient as unknown as Bundle,
        mappings.patientIds,
      );

      expect(actual).toEqual("1234567890");
    });
  });

  describe("Evaluate Patient Race", () => {
    it("should return race category and extension if available", () => {
      const actual = evaluatePatientRace(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("Black or African American\nAfrican");
    });
  });

  describe("Evaluate Patient Ethnicity", () => {
    it("should return ethnicity category and extension if available", () => {
      const actual = evaluatePatientEthnicity(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("Hispanic or Latino\nWhite");
    });
  });

  it("should return tribal affiliation if available", () => {
    const actual = evaluateDemographicsData(
      BundlePatient as unknown as Bundle,
      mappings,
    );
    const ext = actual.availableData.filter(
      (d) => d.title === "Tribal Affiliation",
    );
    expect(ext).toHaveLength(1);
    expect(ext[0].value).toEqual(
      "Fort Mojave Indian Tribe of Arizona, California",
    );
  });

  describe("Evaluate Facility Id", () => {
    it("should return the facility id", () => {
      const actual = evaluateFacilityId(
        BundleEcrMetadata as unknown as Bundle,
        mappings,
      );

      expect(actual).toEqual("112233445566778899");
    });
  });

  describe("Evaluate Encounter ID", () => {
    it("should return the correct Encounter ID", () => {
      const actual = evaluateEncounterId(
        BundleEcrMetadata as unknown as Bundle,
        mappings,
      );

      expect(actual).toEqual("123456789");
    });
  });

  describe("Evaluate Encounter Care Team", () => {
    it("should return the correct Encounter care team", () => {
      const actual = evaluateEncounterCareTeamTable(
        BundleEcrMetadata as unknown as Bundle,
        mappings,
      );

      expect(actual).toMatchSnapshot();
    });
  });

  describe("Evaluate PractitionerRoleReference", () => {
    it("should return the organization and practitioner when practitioner role is found ", () => {
      const actual = evaluatePractitionerRoleReference(
        BundlePractitionerRole as unknown as Bundle,
        mappings,
        "PractitionerRole/b18c20c1-123b-fd12-71cf-9dd0abae8ced",
      );

      expect(actual.organization).toEqual({
        id: "d319a926-0eb3-5847-3b21-db8b778b4f07",
        name: "Mos Eisley Medical Center",
        resourceType: "Organization",
      });

      expect(actual.practitioner).toEqual({
        id: "550b9626-bc9e-7d6b-c5d8-e41c2000ab85",
        name: [
          {
            family: "Interface",
          },
        ],
        resourceType: "Practitioner",
      });
    });
    it("should return undefined organization and practitioner when practitioner role is not found", () => {
      const actual = evaluatePractitionerRoleReference(
        BundlePractitionerRole as unknown as Bundle,
        mappings,
        "unknown",
      );

      expect(actual.organization).toBeUndefined();

      expect(actual.practitioner).toBeUndefined();
    });
  });

  describe("Evaluate Emergency Contact", () => {
    it("should return an emergency contact", () => {
      const BundlePatientAndContact = JSON.parse(
        JSON.stringify(BundlePatient),
      ) as unknown as Bundle;
      const patientIndex = BundlePatientAndContact.entry!.findIndex(
        (entry) => entry.resource?.resourceType === "Patient",
      );

      (
        BundlePatientAndContact.entry![patientIndex].resource as Patient
      ).contact = [
        {
          relationship: [
            {
              coding: [
                {
                  display: "sister",
                },
              ],
            },
          ],
          telecom: [
            {
              system: "phone",
              value: "+1-555-995-9999",
              use: "home",
            },
          ],
          name: {
            given: ["Anastasia", "Bubbletea"],
            family: "Pizza",
          },
          address: {
            use: "home",
            line: ["999 Single Court"],
            city: "BEVERLY HILLS",
            state: "CA",
            country: "USA",
            postalCode: "90210",
            district: "LOS ANGELE",
          },
        },
      ];
      const actual = evaluateEmergencyContact(
        BundlePatientAndContact,
        mappings,
      );
      expect(actual).toEqual(
        `Sister\nAnastasia Bubbletea Pizza\n999 Single Court\nBeverly Hills, CA\n90210, USA\nHome: 555-995-9999`,
      );
    });
    it("should return multiple emergency contacts", () => {
      const BundlePatientAndContact = JSON.parse(
        JSON.stringify(BundlePatient),
      ) as unknown as Bundle;
      const patientIndex = BundlePatientAndContact.entry!.findIndex(
        (entry) => entry.resource?.resourceType === "Patient",
      );

      (
        BundlePatientAndContact.entry![patientIndex].resource as Patient
      ).contact = [
        {
          relationship: [
            {
              coding: [
                {
                  display: "sister",
                },
              ],
            },
          ],
          telecom: [
            {
              system: "phone",
              value: "+1-555-995-9999",
              use: "home",
            },
          ],
          name: {
            given: ["Anastasia", "Bubbletea"],
            family: "Pizza",
          },
          address: {
            use: "home",
            line: ["999 Single Court"],
            city: "BEVERLY HILLS",
            state: "CA",
            country: "USA",
            postalCode: "90210",
            district: "LOS ANGELE",
          },
        },
        {
          relationship: [
            {
              coding: [
                {
                  display: "brother",
                },
              ],
            },
          ],
          name: {
            given: ["Alberto", "Bonanza", "Bartholomew"],
            family: "Eggbert",
          },
          telecom: [
            {
              system: "phone",
              value: "+1-555-995-1000",
              use: "home",
            },
            {
              system: "fax",
              value: "+1-555-995-1001",
              use: "home",
            },
          ],
        },
      ];
      const actual = evaluateEmergencyContact(
        BundlePatientAndContact,
        mappings,
      );
      expect(actual).toEqual(
        `Sister\nAnastasia Bubbletea Pizza\n999 Single Court\nBeverly Hills, CA\n90210, USA\nHome: 555-995-9999\n\nBrother\nAlberto Bonanza Bartholomew Eggbert\nHome: 555-995-1000\nHome Fax: 555-995-1001`,
      );
    });
    it("should not return empty space when address is not available in", () => {
      const BundlePatientAndContact = JSON.parse(
        JSON.stringify(BundlePatient),
      ) as unknown as Bundle;
      const patientIndex = BundlePatientAndContact.entry!.findIndex(
        (entry) => entry.resource?.resourceType === "Patient",
      );

      (
        BundlePatientAndContact.entry![patientIndex].resource as Patient
      ).contact = [
        {
          relationship: [
            {
              coding: [
                {
                  display: "sister",
                },
              ],
            },
          ],
          name: {
            given: ["Anastasia", "Bubbletea"],
            family: "Pizza",
          },
          telecom: [
            {
              system: "phone",
              value: "+1-555-995-9999",
              use: "home",
            },
          ],
        },
      ];
      const actual = evaluateEmergencyContact(
        BundlePatientAndContact,
        mappings,
      );
      expect(actual).toEqual(
        `Sister\nAnastasia Bubbletea Pizza\nHome: 555-995-9999`,
      );
    });
    it("should return undefined if a patient has no contact", () => {
      const actual = evaluateEmergencyContact(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toBeUndefined();
    });
  });

  describe("Evaluate Patient Address", () => {
    it("should return the 1 address", () => {
      const actual = evaluatePatientAddress(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("1 Main St\nCloud City, CA\n00000, US");
    });
    it("should return all 3 of the addresses", () => {
      const actual = evaluatePatientAddress(
        BundlePatientMultiple as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual(
        "Home:\n" +
          "1 Mos Espa\n" +
          "Tatooine, CA\n" +
          "93523-2800, US\n" +
          "\n" +
          "Vacation:\n" +
          "10 Canyon Valley\n" +
          "Ben's Mesa, TN\n" +
          "00047, America\n" +
          "\n" +
          "Work:\n" +
          "1 Main St\n" +
          "Death Star, AZ\n" +
          "00001, USA",
      );
    });
  });

  describe("Evaluate Patient Name", () => {
    it("should return the 1 name", () => {
      const actual = evaluatePatientName(
        BundlePatient as unknown as Bundle,
        mappings,
        false,
      );
      expect(actual).toEqual("Han Solo");
    });
    it("should return all 2 of the names", () => {
      const actual = evaluatePatientName(
        BundlePatientMultiple as unknown as Bundle,
        mappings,
        false,
      );
      expect(actual).toEqual(
        "Official: Anakin Skywalker\n" + "Nickname: Darth Vader",
      );
    });
    it("should only return the official name for the banner", () => {
      const actual = evaluatePatientName(
        BundlePatientMultiple as unknown as Bundle,
        mappings,
        true,
      );
      expect(actual).toEqual("Anakin Skywalker");
    });
    it("should only return the official name for the banner", () => {
      const actual = evaluatePatientName(
        BundlePatient as unknown as Bundle,
        mappings,
        true,
      );
      expect(actual).toEqual("Han Solo");
    });
  });

  describe("Evaluate Alcohol Use", () => {
    it("should return the use, intake comment", () => {
      const actual = evaluateAlcoholUse(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual(
        "Use: Current drinker of alcohol (finding)\n" +
          "Intake (standard drinks/week): .29/d\n" +
          "Comment: 1-2 drinks 2 to 4 times a month",
      );
    });
    it("should empty string because there is no use, intake, or comment", () => {
      const actual = evaluateAlcoholUse(
        BundlePatientMultiple as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("");
    });
  });

  describe("Evaluate Patient Vital Status", () => {
    function getPatientBundle(deceased: boolean) {
      return {
        resourceType: "Bundle",
        entry: [
          {
            resource: {
              resourceType: "Patient",
              deceasedBoolean: deceased,
            },
          },
        ],
      };
    }

    it("should return an empty string when no `deceasedBoolean` value is present", () => {
      const actual = evaluatePatientVitalStatus(
        BundlePatient as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("");
    });

    it("should return `Alive` when `deceasedBoolean` is `false`", () => {
      const actual = evaluatePatientVitalStatus(
        getPatientBundle(false) as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("Alive");
    });

    it("should return `Deceased` when `deceasedBoolean` is `true`", () => {
      const actual = evaluatePatientVitalStatus(
        getPatientBundle(true) as unknown as Bundle,
        mappings,
      );
      expect(actual).toEqual("Deceased");
    });
  });

  describe("Evaluate Patient language", () => {
    it("Should display language, proficiency, and mode", () => {
      const actual = evaluatePatientLanguage(
        BundlePatient as unknown as Bundle,
        mappings,
      );

      expect(actual).toEqual("English\nGood\nExpressed spoken");
    });

    it("Should only display preferred languages", () => {
      const patient = {
        resourceType: "Bundle",
        entry: [
          {
            resource: {
              resourceType: "Patient",
              communication: [
                {
                  language: {
                    coding: [
                      {
                        system: "urn:ietf:bcp:47",
                        code: "es",
                        display: "Spanish",
                      },
                    ],
                  },
                },
                {
                  preferred: true,
                  language: {
                    coding: [
                      {
                        system: "urn:ietf:bcp:47",
                        code: "en",
                        display: "English",
                      },
                    ],
                  },
                },
                {
                  preferred: true,
                  language: {
                    coding: [
                      {
                        system: "urn:ietf:bcp:47",
                        code: "hi",
                        display: "Hindi",
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      };

      const actual = evaluatePatientLanguage(
        patient as unknown as Bundle,
        mappings,
      );

      expect(actual).toEqual("English\n\nHindi");
    });
    it("Should display language when there are no preferred languages", () => {
      const patient = {
        resourceType: "Bundle",
        entry: [
          {
            resource: {
              resourceType: "Patient",
              communication: [
                {
                  language: {
                    coding: [
                      {
                        system: "urn:ietf:bcp:47",
                        code: "es",
                        display: "Spanish",
                      },
                    ],
                  },
                },
                {
                  language: {
                    coding: [
                      {
                        system: "urn:ietf:bcp:47",
                        code: "en",
                        display: "English",
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      };

      const actual = evaluatePatientLanguage(
        patient as unknown as Bundle,
        mappings,
      );

      expect(actual).toEqual("Spanish\n\nEnglish");
    });
  });

  describe("Get Human Readable CodeableConcept", () => {
    it("should return undefined if no coding is available", () => {
      const codeableConcept = undefined;

      const actual = getHumanReadableCodeableConcept(codeableConcept);

      expect(actual).toBeUndefined();
    });

    it("should return the text value if available", () => {
      const textValue = "this is condition";
      const codeableConcept: CodeableConcept = {
        text: textValue,
        coding: [
          {
            display: "Condition",
            code: "64572001",
          },
        ],
      };

      const actual = getHumanReadableCodeableConcept(codeableConcept);
      expect(actual).toEqual(textValue);
    });

    it("should return the first display value if there is no text value", () => {
      const correctDisplayValue = "Condition";
      const codeableConcept: CodeableConcept = {
        coding: [
          {
            display: "Condition",
            code: "64572001",
          },
          {
            display: "A Condition",
            code: "AC",
          },
        ],
      };

      const actual = getHumanReadableCodeableConcept(codeableConcept);
      expect(actual).toEqual(correctDisplayValue);
    });

    it("should return the code and system of the first coding with both of them if there is no text or display value", () => {
      const codeValue = "64572001";
      const systemValue = "http://snomed.info/sct";
      const codeableConcept: CodeableConcept = {
        coding: [
          {
            code: "AC",
          },
          {
            code: codeValue,
            system: systemValue,
          },
        ],
      };

      const actual = getHumanReadableCodeableConcept(codeableConcept);
      expect(actual).toEqual(`${codeValue} (${systemValue})`);
    });

    it("should return the code of the first first coding with a code if there is no text, display, or a code/system pair", () => {
      const codeValue = "64572001";
      const codeableConcept: CodeableConcept = {
        coding: [
          {
            code: codeValue,
          },
        ],
      };

      const actual = getHumanReadableCodeableConcept(codeableConcept);
      expect(actual).toEqual(codeValue);
    });
  });
});
