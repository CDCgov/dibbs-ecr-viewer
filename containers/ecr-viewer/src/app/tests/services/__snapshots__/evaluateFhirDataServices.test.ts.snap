// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluateFhirDataServices tests Evaluate Encounter Care Team should return the correct Encounter care team 1`] = `
<JsonTable
  className="caption-data-title margin-y-0"
  outerBorder={true}
  rawTable={
    {
      "resultName": "Encounter Care Team",
      "tables": [
        [
          {
            "Dates": {
              "value": <span
                className="no-data text-italic text-base"
              >
                No data
              </span>,
            },
            "Name": {
              "value": <span
                className="no-data text-italic text-base"
              >
                No data
              </span>,
            },
            "Role": {
              "value": "ATND",
            },
          },
        ],
      ],
    }
  }
/>
`;
